gui/wxpython/gui_core/menu.py

```diff
@@ -305,3 +308,56 @@ def OnItemSelected(self, node):
             label = data['description']
 
         self.showNotification.emit(message=label)
+
+
+class FileHistory:
+    """Add recent files history menu"""
+
+    file_history = '.file_history'
+
+    def __init__(
+            self, app_name, parent_menu, pos,
+            history_len=10, text_widget=None,
+    ):
+        self.text_widget = text_widget
+        self.filehistory = wx.FileHistory(maxFiles=history_len)
+        self.config = wx.FileConfig(
+            style=wx.CONFIG_USE_LOCAL_FILE,
+            localFilename=os.path.join(
+                utils.GetSettingsPath(), self.file_history,
+            ),
+        )
+        self.config.SetPath(strPath=app_name)
+        self.filehistory.Load(self.config)
+
+        self.recent = wx.Menu()
+        self.filehistory.UseMenu(self.recent)
+        self.filehistory.AddFilesToMenu()
+
+        parent_menu.Insert(
+            pos=pos, id=wx.ID_ANY, text=_('&Recent Files'),
+            submenu=self.recent,
+        )
+
+        self.recent.Bind(
+            wx.EVT_MENU_RANGE, self.OnFileHistory,
+            id=wx.ID_FILE1, id2=wx.ID_FILE + history_len,
+        )
+
+    def OnFileHistory(self, event):
+        file_num = event.GetId() - wx.ID_FILE1
+        self.path = self.filehistory.GetHistoryFile(file_num)
+        self.filehistory.AddFileToHistory(self.path)  # move up the list
+
+        if self.text_widget is not None:
+            self.SetText()
```

Comment:
This, `text_widget`, and `def SetText` should not be here. This is something to handle ideally in the controller or the frame if really needed.

You can solve this by callback, wx event, or signal. Plain callbacks are fine, but are not the primary method in the wxGUI/wxPython (although certainly used a lot). Wx events are kind of cumbersome esp. when you define your own. Signals are quite nice though. Almost as simple as callbacks, but in many common cases as powerful as wx events. Search for `Signal(` in the wxGUI code base for examples. The library doc for them should be pretty good too.

You could try to sketch how this would be used for workspaces which can help with writing a general implementation for this class.

Reply:
Yes I understand I already fixed it.